java/arcs/android/util/ProtoPrefetcher.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.android.util
+
+import android.os.Trace
+import arcs.android.crdt.CrdtCountProto
+import arcs.android.crdt.CrdtDataProto
+import arcs.android.crdt.CrdtEntityProto
+import arcs.android.crdt.CrdtEntityReferenceProto
+import arcs.android.crdt.CrdtOperationProto
+import arcs.android.crdt.CrdtSetProto
+import arcs.android.crdt.CrdtSingletonProto
+import arcs.android.crdt.RawEntityProto
+import arcs.android.crdt.ReferencablePrimitiveProto
+import arcs.android.crdt.ReferencableProto
+import arcs.android.crdt.ReferencableSetProto
+import arcs.android.crdt.ReferenceProto
+import arcs.android.crdt.VersionMapProto
+import arcs.android.storage.ProxyMessageProto
+import java.util.concurrent.Executors
+
+/**
+ * Run asynchronously as a cache warmer to suffer from Arcs protos initialization
+ * overhead as well as cache misses on behalf of whoever needs Arcs capabilities.
+ *
+ * The initialization overhead typically comes from instantiating default instance
+ * and parsing/building proto schema.
+ *
+ * Recommend executing [ProtoPrefetcher.prefetch] as early as possible i.e.
+ * the entry of Application lifecycle: `onCreate()`
+ */
+object ProtoPrefetcher {
+    /** Identify [ProtoPrefetcher] in thread lists and system traces. */
+    private const val name = "ProtoPrefetcher"
+
+    /**
+     * Single-threaded executor.
+     *
+     * TODO(ianchang): tune background priority if needed.
+     */
+    private val executor = Executors.newFixedThreadPool(1) {
+        Thread(it).apply { name = this@ProtoPrefetcher.name }
+    }
```

Comment:
Curious, what makes this better than using `Executors.newSingleThreadExecutor()`?

Reply:
Using newSingleThreadExecutor is 99% identical to using newFixedThreadPool(1) except the latter allows "reconfiguration" and the former is wrapped by a delegated finalized class.

I was thinking the "prefetch" idea can be extended to broader usage and dynamically boost it with higher parallelism by executing multiple tasks and temporarily executor.maximumPoolSize = n to expedite prefetching process then reset executor.maximumPoolSize = 1 after finishing the prefetch tasks to force the numbers of standby threads down to 1.